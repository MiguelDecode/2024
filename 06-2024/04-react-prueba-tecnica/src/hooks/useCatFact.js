import { useEffect, useState } from 'react'
import { getRandomFact } from '../services/getRandomFact'

export const useCatFact = () => {
  const [fact, setFact] = useState()
  const [factError, setFactError] = useState()

  const refreshFact = () => {
    getRandomFact().then(({ fact, err }) => {
      setFact(fact)
      setFactError(err)
    })
  }

  useEffect(refreshFact, [])

  return { fact, factError, refreshFact }
}
