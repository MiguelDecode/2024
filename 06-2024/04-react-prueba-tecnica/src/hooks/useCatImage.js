import { useEffect, useState } from 'react'
import { getImageByWords } from '../services/getImageByWords'

export const useCatImage = ({ fact }) => {
  const [imageUrl, setImageUrl] = useState(null)
  const [imageError, setImageError] = useState()

  useEffect(() => {
    if (!fact) return

    const threeFirstWords = fact.split(' ', 3)

    getImageByWords(threeFirstWords).then(({ url, err }) => {
      setImageUrl(url)
      setImageError(err)
    })
  }, [fact])

  return { imageUrl, imageError }
}
