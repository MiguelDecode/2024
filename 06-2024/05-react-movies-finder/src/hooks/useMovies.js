import { useRef, useState, useMemo, useCallback } from 'react'
import { searchMovies } from '../services/movies'

export const useMovies = ({ search, sort }) => {
  const [movies, setMovies] = useState([])
  const [loading, setLoading] = useState(false)
  const [error, setError] = useState(null)
  const previousSearch = useRef(search)

  const formatYear = (year) => Number(year.slice(0, 4))

  const getMovies = useCallback(async ({ search }) => {
    if (search === previousSearch.current) return

    try {
      setLoading(true)
      setError(null)
      previousSearch.current = search
      const newMovies = await searchMovies({ search })
      setMovies(newMovies)
    } catch (err) {
      setError(err.message)
    } finally {
      setLoading(false)
    }
  }
  , [])

  const sortedMovies = useMemo(() => {
    return sort
      ? [...movies].sort((a, b) => formatYear(a.year) - formatYear(b.year))
      : movies
  }, [sort, movies])

  return { movies: sortedMovies, loading, error, getMovies }
}
