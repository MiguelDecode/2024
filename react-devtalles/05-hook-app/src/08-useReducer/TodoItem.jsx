import PropTypes from "prop-types";

export const TodoItem = ({ todo, removeTodo, changeToggleTodo }) => {
  return (
    <li className="list-group-item d-flex justify-content-between">
      <span
        aria-label="span"
        className={`align-self-center ${
          todo.done ? "text-decoration-line-through" : ""
        }`}
        onDoubleClick={() => changeToggleTodo(todo.id)}
      >
        {todo.desc}
      </span>
      <button className="btn btn-danger" onClick={() => removeTodo(todo.id)}>
        Borrar
      </button>
    </li>
  );
};

TodoItem.propTypes = {
  todo: PropTypes.object,
  removeTodo: PropTypes.func,
  changeToggleTodo: PropTypes.func,
};
