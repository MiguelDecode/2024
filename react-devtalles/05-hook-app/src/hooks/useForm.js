import { useState } from "react";

export const useForm = (initialForm = {}) => {
  const [formState, setFormState] = useState(initialForm);


  const handleInputChange = (event) => {
    const key = event.target.name;
    const value = event.target.value;

    setFormState({ ...formState, [key]: value });
  };

  const handleReset = () => setFormState(initialForm);

  return {
    formState,
    handleInputChange,
    onResetForm: handleReset,
  };
};
