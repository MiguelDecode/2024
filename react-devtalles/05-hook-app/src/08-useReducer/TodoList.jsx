import PropTypes from "prop-types";
import { TodoItem } from "./TodoItem";

export const TodoList = ({ todos = [], removeTodo, changeToggleTodo }) => {
  return (
    <ul className="list-group">
      {todos.map((todo) => (
        <TodoItem
          key={todo.id}
          todo={todo}
          removeTodo={removeTodo}
          changeToggleTodo={changeToggleTodo}
        />
      ))}
    </ul>
  );
};

TodoList.propTypes = {
  todos: PropTypes.array,
  removeTodo: PropTypes.func,
  changeToggleTodo: PropTypes.func,
};
