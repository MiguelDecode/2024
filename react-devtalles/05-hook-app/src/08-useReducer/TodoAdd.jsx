import PropTypes from "prop-types";
import { useForm } from "../hooks/useForm";

export const TodoAdd = ({ addNewTodo }) => {
  const { formState, handleInputChange, onResetForm } = useForm({
    desc: "",
  });

  const handleSubmit = (event) => {
    event.preventDefault();

    if (formState.length <= 1) return;

    const newTodo = {
      id: crypto.randomUUID(),
      desc: formState.desc,
      done: false,
    };

    addNewTodo(newTodo);
    onResetForm();
  };

  return (
    <form onSubmit={handleSubmit}>
      <input
        type="text"
        placeholder="Que hay que hacer?"
        className="form-control"
        name="desc"
        value={formState.desc}
        onChange={handleInputChange}
      />

      <button type="submit" className="btn btn-outline-primary mt-2">
        Agregar
      </button>
    </form>
  );
};

TodoAdd.propTypes = {
  addNewTodo: PropTypes.func,
};
