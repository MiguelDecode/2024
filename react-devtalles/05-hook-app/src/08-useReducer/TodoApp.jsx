import { useEffect } from "react";
import { TodoList } from "./TodoList";
import { TodoAdd } from "./TodoAdd";
import { useTodo } from "../hooks/useTodo";

export const TodoApp = () => {
  const {
    todos,
    totalTodos,
    pendingTodos,
    addNewTodo,
    removeTodo,
    changeToggleTodo,
  } = useTodo();

  useEffect(() => {
    localStorage.setItem("todos", JSON.stringify(todos) || []);
  }, [todos]);

  return (
    <>
      <h1>
        TodoApp: {totalTodos}, <small>pendientes: {pendingTodos}</small>
      </h1>

      <hr />

      <div className="row">
        <div className="col-7">
          <TodoList
            todos={todos}
            removeTodo={removeTodo}
            changeToggleTodo={changeToggleTodo}
          />
        </div>

        <div className="col-5">
          <h4>Agredar TODO</h4>

          <hr />

          <TodoAdd addNewTodo={addNewTodo} />
        </div>
      </div>
    </>
  );
};
