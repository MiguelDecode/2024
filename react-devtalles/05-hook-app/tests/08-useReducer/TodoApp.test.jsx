import { render, screen } from "@testing-library/react";
import { TodoApp } from "../../src/08-useReducer/TodoApp";
import { useTodo } from "../../src/hooks/useTodo";

jest.mock("../../src/hooks/useTodo.js");

describe("Pruebas en TodoApp", () => {
  useTodo.mockReturnValue({
    todos: [
      { id: 1, desc: "Todo #1", done: false },
      { id: 2, desc: "Todo #2", done: true },
    ],
    totalTodos: 2,
    pendingTodos: 1,
    addNewTodo: jest.fn(),
    removeTodo: jest.fn(),
    changeToggleTodo: jest.fn(),
  });

  test("Debe de hacer match con el snapshot", () => {
    const { container } = render(<TodoApp />);

    expect(container).toMatchSnapshot();
  });

  test("Debe de mostar el componente correctamente", () => {
    render(<TodoApp />);

    expect(screen.getByText("Todo #1")).toBeTruthy();
    expect(screen.getByText("Todo #2")).toBeTruthy();
    expect(screen.getByRole("textbox")).toBeTruthy();
  });
});
