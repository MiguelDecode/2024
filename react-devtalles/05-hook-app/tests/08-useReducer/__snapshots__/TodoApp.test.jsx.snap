// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en TodoApp Debe de hacer match con el snapshot 1`] = `
<div>
  <h1>
    TodoApp: 
    2
    , 
    <small>
      pendientes: 
      1
    </small>
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-7"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex justify-content-between"
        >
          <span
            aria-label="span"
            class="align-self-center "
          >
            Todo #1
          </span>
          <button
            class="btn btn-danger"
          >
            Borrar
          </button>
        </li>
        <li
          class="list-group-item d-flex justify-content-between"
        >
          <span
            aria-label="span"
            class="align-self-center text-decoration-line-through"
          >
            Todo #2
          </span>
          <button
            class="btn btn-danger"
          >
            Borrar
          </button>
        </li>
      </ul>
    </div>
    <div
      class="col-5"
    >
      <h4>
        Agredar TODO
      </h4>
      <hr />
      <form>
        <input
          class="form-control"
          name="desc"
          placeholder="Que hay que hacer?"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-2"
          type="submit"
        >
          Agregar
        </button>
      </form>
    </div>
  </div>
</div>
`;
